src/main/cpp/src/datetime_parser.cu

```diff
@@ -0,0 +1,453 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <map>
+#include <numeric>
+#include <vector>
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/utilities/vector_factories.hpp>
+#include <cudf/strings/convert/convert_datetime.hpp>
+#include <cudf/strings/detail/converters.hpp>
+#include <cudf/strings/detail/strings_children.cuh>
+#include <cudf/strings/detail/utilities.cuh>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/span.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/wrappers/timestamps.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+#include <thrust/execution_policy.h>
+#include <thrust/for_each.h>
+#include <thrust/functional.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/logical.h>
+#include <thrust/optional.h>
+#include <thrust/pair.h>
+#include <thrust/transform.h>
+
+#include "datetime_parser.hpp"
+
+namespace {
+
+using timestamp_components = spark_rapids_jni::timestamp_components;
+
+/**
+ * Get the timestamp from epoch from a local date time in a specific time zone.
+ * Note: local date time may be overlap or gap, refer to `ZonedDateTime.of`
+ *
+ */
+__device__ cudf::timestamp_us
+create_timestamp_from_components_and_zone(timestamp_components local_timestamp_components,
+                                          cudf::string_view time_zone) {
+  // TODO: implements:
+  //   val localDateTime = LocalDateTime.of(localDate, localTime)
+  //   val zonedDateTime = ZonedDateTime.of(localDateTime, zoneId)
+  //   val instant = Instant.from(zonedDateTime) // main work
+  //   instantToMicros(instant)
+  // here just return a zero
+  return cudf::timestamp_us{cudf::duration_us{0L}};
+}
+
+__device__ __host__ inline bool is_digit(const char chr) {
+  return (chr >= '0' && chr <= '9');
+}
+
+__device__ __host__ inline bool is_whitespace(const char chr) {
+  switch (chr) {
+    case ' ':
+    case '\r':
+    case '\t':
+    case '\n': return true;
+    default: return false;
+  }
+}
+
+/**
+ * first trim the time zone,
+ * then format (+|-)h:mm, (+|-)hh:m or (+|-)h:m to (+|-)hh:mm
+ * Refer to: https://github.com/apache/spark/blob/v3.5.0/sql/api/src/main/scala/
+ * org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala#L39
+ */
+__device__ __host__ cudf::string_view format_zone_id(const cudf::string_view &time_zone_id) {
```

Comment:
nit: can we mention the supported formats in code comments as well?    
```
   * `[+-]yyyy*`
   * `[+-]yyyy*-[m]m`
   * `[+-]yyyy*-[m]m-[d]d`
   * `[+-]yyyy*-[m]m-[d]d `
   * `[+-]yyyy*-[m]m-[d]d [h]h:[m]m:[s]s.[ms][ms][ms][us][us][us][zone_id]`
   * `[+-]yyyy*-[m]m-[d]dT[h]h:[m]m:[s]s.[ms][ms][ms][us][us][us][zone_id]`
   * `[h]h:[m]m:[s]s.[ms][ms][ms][us][us][us][zone_id]`
   * `T[h]h:[m]m:[s]s.[ms][ms][ms][us][us][us][zone_id]`
```

Reply:
Updated, refer to the header file: datetime_parser.hpp.